[red-knot] Use ternary decision diagrams (TDDs) for visibility constraints

dcreager opened the pull request:
We now use ternary decision diagrams (TDDs) to represent visibility constraints.  A TDD is just like a BDD ([_binary_ decision diagram](https://en.wikipedia.org/wiki/Binary_decision_diagram)), but with "ambiguous" as an additional allowed value.  Unlike the previous representation, TDDs are strongly normalizing, so equivalent ternary formulas are represented by exactly the same graph node, and can be compared for equality in constant time.

We currently have a slight 1-3% performance regression with this in place, according to local testing.  However, we also have a _5× increase_ in performance for pathological cases, since we can now remove the recursion limit when we evaluate visibility constraints.

As follow-on work, we are now closer to being able to remove the `simplify_visibility_constraint` calls in the semantic index builder.  In the vast majority of cases, we now see (for instance) that the visibility constraint after an `if` statement, for bindings of symbols that weren't rebound in any branch, simplifies back to `true`.  But there are still some cases we generate constraints that are cyclic.  With fixed-point cycle support in salsa, or with some careful analysis of the still-failing cases, we might be able to remove those.

carljm commented:
Very excited about this!

Per your request (and the "draft" status), I'll refrain from a detailed look at the code until you declare it ready for review. Just one general thought, based on the title of the PR: when I looked at this previously, my conclusion was that we don't actually need TDDs here, BDDs suffice. If we build a BDD of the constraints, and then evaluate that BDD, anytime a decision node evaluates to "ambiguous", we can immediately short-circuit to a result of "ambiguous". I guess another way to say this is that a TDD where the "ambiguous" exit of every node leads immediately to the "ambiguous" terminal (which is our scenario, I think), is isomorphic to a BDD where the "ambiguous" exits can be implied.

Just mentioning that in case it's useful or allows any simplification (I don't know if it does, since I haven't looked at your code yet.). Ultimately I'll be happy with any solution that performs well and allows us to get rid of the arbitrary depth limit, no matter what we call it!

(One other thought: if we are getting rid of the depth limit, we should run this on Black and verify that the big perf regression @sharkdp observed is in fact gone.)

dcreager commented:
> If we build a BDD of the constraints, and then evaluate that BDD, anytime a decision node evaluates to "ambiguous", we can immediately short-circuit to a result of "ambiguous".

Hmm, I thought we'd need to model ambiguous edges fully, since (e.g.) `ambiguous ∧ false = false`.  So if we have a constraint of `C1 ∧ C2`, where `C1` is `amb` and `C2` is `false`, I think your optimization would over-approximate the result to `amb` instead of `false`.  Am I following that right?

> (One other thought: if we are getting rid of the depth limit, we should run this on Black and verify that the big perf regression @sharkdp observed is in fact gone.)

Ah, that makes me a bit happier about the performance numbers I'm seeing!  Codspeed is reporting a 3% regression with this in place.  When I test locally, using hyperfine on black, I see similar numbers.  But that might not be apples-to-apples, since `main` has a recursion limit in place but this feature branch does not.

carljm commented:
> I thought we'd need to model ambiguous edges fully, since (e.g.) `ambiguous ∧ false = false`

You're right, of course! I remembered that I'd reached the conclusion we could do this with just BDDs, but when I tried to recall the reasoning for that conclusion, I thought I'd come up with a simpler reason, but instead it was just a wrong reason :)

The actual reason I'd reached this conclusion before was because OxiDD will allow you to supply "resolutions" (that is, t/f values) for nodes in a BDD and will then simplify the BDD accordingly. So I thought if we built a BDD and then resolved the nodes that we can infer a statically-known truthiness for, the BDD would either simplify to a terminal, or not, in which case it remains ambiguous.

But I don't know if that's actually a good/efficient approach, compared to building a TDD and evaluating it!

> Codspeed is reporting a 3% regression with this in place. When I test locally, using hyperfine on black, I see similar numbers. But that might not be apples-to-apples,

IIRC @sharkdp was seeing like 5x regression on Black without the recursion limit, so it seems like you're doing much better on the pathological cases. It's too bad if this comes with an across-the-board regression, but I think it's worth it to get rid of the arbitrary limit. We probably will need such limits in some places, but they are a really bad experience when users encounter them, so as much as we can avoid them, or make them so high that users will never encounter them in realistic code, we should try to do so.

Of course we should also take a look at profiles and see if there are ways to bring down the regression here!

dcreager commented:
I've extracted the refactoring parts of this into a separate PR to make everything easier to review: https://github.com/astral-sh/ruff/pull/15913

dcreager commented:
> IIRC @sharkdp was seeing like 5x regression on Black without the recursion limit, so it seems like you're doing much better on the pathological cases. It's too bad if this comes with an across-the-board regression, but I think it's worth it to get rid of the arbitrary limit. We probably will need such limits in some places, but they are a really bad experience when users encounter them, so as much as we can avoid them, or make them so high that users will never encounter them in realistic code, we should try to do so.
> 
> Of course we should also take a look at profiles and see if there are ways to bring down the regression here!

As expected, the execution time shifts from being mostly in the evaluate step on `main`:

![main](https://github.com/user-attachments/assets/9985524d-9ec6-496f-a0aa-ecedcaaf4474)

to being mostly in the build step with this PR:

![feature](https://github.com/user-attachments/assets/2d2657a4-69b2-4107-9ea7-60323a08b3cc)

dcreager commented:
I think this is at a good point now.  In local testing, I've gotten the performance regression down to 1-2%.  Interested to see what codspeed says about the latest commits on the branch.

carljm commented:
For some reason CodSpeed thinks it doesn't have data on the baseline for this PR, but just comparing the results here with the ones on main (or `vc-api`), it looks like about 2% regression (89.5 ms -> 91.6 ms).

carljm commented:
> As follow-on work, we are now closer to being able to remove the simplify_visibility_constraint calls in the semantic index builder. In the vast majority of cases, we now see (for instance) that the visibility constraint after an if statement, for bindings of symbols that weren't rebound in any branch, simplifies back to true. But there are still some cases we generate constraints that are cyclic. With fixed-point cycle support in salsa, or with some careful analysis of the still-failing cases, we might be able to remove those.

I would definitely like to see the failing cases and understand why they need to be cyclic, or if there's some missing simplification (other than the "no new bindings" heuristic that doesn't work with with terminals) that could avoid the cycles. But this can be a follow-up.

dcreager commented:
> For some reason CodSpeed [thinks](https://codspeed.io/astral-sh/ruff/branches/dcreager%2Ftdd) it doesn't have data on the baseline for this PR, but just comparing the results here with the ones on main (or `vc-api`), it looks like about 2% regression (89.5 ms -> 91.6 ms).

I rebased after merging #15913, which should hopefully update codspeed

dcreager commented:
> I would definitely like to see the failing cases and understand why they need to be cyclic, or if there's some missing simplification (other than the "no new bindings" heuristic that doesn't work with with terminals) that could avoid the cycles. But this can be a follow-up.

As one example, [`flake8_return/RET503.py`](https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/resources/test/fixtures/flake8_return/RET503.py) works fine when analyzed as a `.py` file, but has a cycle when analyzed as a `.pyi` file

dcreager commented:
> As one example, [`flake8_return/RET503.py`](https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/resources/test/fixtures/flake8_return/RET503.py) works fine when analyzed as a `.py` file, but has a cycle when analyzed as a `.pyi` file

And I minimized it down to this repro:

```py
def foo() -> int:
    def baz() -> int:
        return 1
    if baz() > 3:
        return 1
    raise ValueError
```

Passes as a `.py` file, cycles as a `.pyi` file. Passes if you remove the `if` statement.

carljm commented:
Summarizing from Discord: I think a) such code would never actually occur in a stub file; they shouldn't even have function bodies other than ellipsis (and perhaps we should make this explicit in how we handle them), and b) it would be fine to go ahead and remove the simplify calls and mark that file xfail (though ideal if we could mark only the pyi version of it as xfail?) pending fixpoint iteration, which I think would solve the panic.

carljm commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,392 @@ const MAX_RECURSION_DEPTH: usize = 24;
+
+    /// Adds an interior node, ensuring that we always use the same visibility constraint ID for
+    /// equal nodes.
+    #[allow(clippy::cast_possible_truncation)]
```
Is it possible to silence this on just the one line, instead of the whole function?

dcreager replied:
These went away with the switch to `IndexVec`

carljm reviewed (approved):
Looks great to me!

Micha or David may have Rust/perf suggestions I missed, but I think it's OK to go ahead with this and handle any such comments as follow-ups.

dcreager commented:
> Summarizing from Discord: I think a) such code would never actually occur in a stub file; they shouldn't even have function bodies other than ellipsis (and perhaps we should make this explicit in how we handle them), and b) it would be fine to go ahead and remove the simplify calls and mark that file xfail (though ideal if we could mark only the pyi version of it as xfail?) pending fixpoint iteration, which I think would solve the panic.

Done.  Performance between keeping and removing the simplify calls seems to be a wash.  I've removed them in the interests of simpler code, and added a pyi-only xfail for the one failing test.

MichaReiser commented:
As I haven't reviewed the code yet but the regression might simply come from that we now evaluate (or simplify) more constraints than before because the semantic indexer processes all constraints whereas simplification happened lazily before (and only for the queried definitions)

dcreager commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,392 @@ const MAX_RECURSION_DEPTH: usize = 24;
-            ]),
-        }
-    }
+    constraints: Vec<Constraint<'db>>,
+    interiors: Vec<InteriorNode>,
```
Mostly because I thought it was overkill to add `newtype_index`es for the two vectors, but actually that type safety seems like a good thing.  I'll change it

dcreager commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,392 @@ const MAX_RECURSION_DEPTH: usize = 24;
+    #[allow(clippy::cast_possible_truncation)]
+    fn add_constraint(&mut self, constraint: Constraint<'db>, copy: u8) -> Atom {
+        let index = *self.constraint_cache.entry(constraint).or_insert_with(|| {
+            let index = self.constraints.len() as u32;
```
There's a `debug_assert` that we're less than 2^24 in `from_index_and_copy`, though that happens after the cast.  Switching to `IndexVec` gives us that u32 check for free.

carljm commented:
Hmm, the switch to `IndexVec` (or at least, something in the recent commits) seems to have grown the regression; now we're seeing a significant regression even in incremental check, which is a bit strange.

carljm reviewed (approved):
Semantics and code here look good to me. May be worth understanding why the regression seems to have grown again, and see if that reproduces locally on a larger codebase like Black.

dcreager commented:
> I'm going to revert that commit and add a note about why we're using `Vec`. (Alternatively I could add an `unchecked` constructor for the ID types.)

Ackshually I can write a custom `Idx` impl for a couple of types and bypass the asserts, I think.

dcreager commented:
Still seeing the larger regression with custom `Idx` impls, and again with a revert of the `Vec` → `IndexVec` commit.  I'm stumped, will look at this some more in the morning.

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = f.debug_tuple("ScopedVisibilityConstraintId");
+        match *self {
+            ALWAYS_TRUE => f.field(&format_args!("AlwaysTrue")).finish(),
+            AMBIGUOUS => f.field(&format_args!("Ambiguous")).finish(),
+            ALWAYS_FALSE => f.field(&format_args!("AlwaysFalse")).finish(),
```
```suggestion
            ALWAYS_TRUE => f.field(&"AlwaysTrue").finish(),
            AMBIGUOUS => f.field(&"Ambiguous").finish(),
            ALWAYS_FALSE => f.field(&"AlwaysFalse").finish(),
```

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
+
+impl std::fmt::Debug for ScopedVisibilityConstraintId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = f.debug_tuple("ScopedVisibilityConstraintId");
+        match *self {
+            ALWAYS_TRUE => f.field(&format_args!("AlwaysTrue")).finish(),
+            AMBIGUOUS => f.field(&format_args!("Ambiguous")).finish(),
+            ALWAYS_FALSE => f.field(&format_args!("AlwaysFalse")).finish(),
+            _ => f.field(&self.0).finish(),
+        }
```
Nit: Makes it clearer what's different/shared between the branches:

```suggestion
        let mut f = f.debug_tuple("ScopedVisibilityConstraintId");
        match *self {
            ALWAYS_TRUE => f.field(&"AlwaysTrue"),
            AMBIGUOUS => f.field(&"Ambiguous"),
            ALWAYS_FALSE => f.field(&"AlwaysFalse"),
            _ => f.field(&self.0),
        }
        
        f.finish()
```

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
+// _Atoms_ are the underlying Constraints, which are the variables that are evaluated by the
+// ternary function.
+//
+// _Interior nodes_ provide the TDD structure for the formula. Interior nodes are stored in an
```
I'm not familiar with TDD myself (I'm familiar with test driven development but that's not it). Would it make sense to extend the module level documentation and also include a link to what TDD means in this context?

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
+/// A "variable" that is evaluated as part of a TDD ternary function. For visibility constraints,
+/// this is (one of the copies of) a `Constraint` that represents some runtime property of the
+/// Python code that we are evaluating. We intern these constraints in an arena
+/// ([`VisibilityConstraints::constraints`]). An atom consists of an index into this arena, and a
+/// copy number.
```
It's unclear to me what a copy number is. Can you expand the comment to cover this in more detail. It may also be worthwhile to explain the internal representation. How many constraints can be expressed etc.

dcreager replied:
Done

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
+const ALWAYS_TRUE: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::ALWAYS_TRUE;
+const AMBIGUOUS: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::AMBIGUOUS;
+const ALWAYS_FALSE: ScopedVisibilityConstraintId = ScopedVisibilityConstraintId::ALWAYS_FALSE;
+const SMALLEST_TERMINAL: u32 = ALWAYS_FALSE.0;
```
I'd expected this to also be defined on `ScopedVisiblityConstraitId` as it is used in `is_terminal` and that there's only an alias here.

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
-    }
+    constraints: Vec<Constraint<'db>>,
+    interiors: Vec<InteriorNode>,
+    constraint_cache: FxHashMap<Constraint<'db>, u32>,
+    interior_cache: FxHashMap<InteriorNode, u32>,
```
if we end up not using `IndexVec`, I still recommend to create a newtype wrapper around `u32` to signify what the `u32` here means (it's an index into the constraints/interiors vec?)

dcreager replied:
We're back to using `IndexVec`

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -182,211 +176,389 @@ const MAX_RECURSION_DEPTH: usize = 24;
-            ]),
-        }
-    }
+    constraints: Vec<Constraint<'db>>,
+    interiors: Vec<InteriorNode>,
```
This struct defines a fair amount of heap-allocated data structures and requires a lot of hashing. It would be nice not to have to do that, but I don't see how we can restructure the code to avoid it (unless maybe combining some maps and encoding the `or`/`and` in the key but that might as well turn out to be slower if it happens that the map has to resize more often because of it.

dcreager replied:
The hashing is an important part of the data structure, since we have to intern all of the nodes to guarantee that the TDD is reduced.  (I included a description of what it means to be reduced, and why that's important, in the module documentation that I added).

MichaReiser reviewed (approved):
Nice :) I only have nit/documentation comments. 

It would be nice if we could use `IndexVec` or a newtype wrapper and implement `std::ops::Index` but it seems you tried that and it regressed performance.

MichaReiser commented on crates/red_knot_python_semantic/src/visibility_constraints.rs:
```diff
@@ -285,6 +282,21 @@ impl ScopedVisibilityConstraintId {
+impl Idx for ScopedVisibilityConstraintId {
+    #[inline]
+    fn new(value: usize) -> Self {
+        assert!(value <= (SMALLEST_TERMINAL as usize));
```
Does it make any difference if you remove the assert here just to make sure we compare apples with apples
